airgun/entities/domain.py

```diff
@@ -0,0 +1,192 @@
+    _DEFAULT_ORG = 'Default Organization'
+    _DEFAULT_LOC = 'Default Location'
+    _CREATE_SUCCESSFUL = 'Successfully created {}.'
+    _UPDATE_SUCCESSFUL = 'Successfully updated {}.'
+    _DELETE_SUCCESSFUL = 'Successfully deleted {}.'
+
+    @staticmethod
+    def _submit_domain_form(view, **kwargs):
+        """Create or update domain.
+
+        Form is the same for both cases, we use 1 method here to fill the form.
+        """
+        dns_domain = kwargs.get('dns_domain')
+        full_name = kwargs.get('full_name')
+        dns_capsule = kwargs.get('dns_capsule')
+        parameters = kwargs.get('parameters')
+        locations = kwargs.get('locations')
+        organizations = kwargs.get('organizations')
+
+        # Fill 'Domain' tab
+        if dns_domain:
+            view.domain.dns_domain.fill(dns_domain)
+        if full_name:
+            view.domain.full_name.fill(full_name)
+        if dns_capsule:
+            view.domain.dns_capsule.fill(dns_capsule)
+
+        # Fill 'Parameters' tab
+        if parameters:
+            params = []
+            for param_name, param_value in parameters.items():
+                params.append({'name': param_name, 'value': param_value})
+            view.parameters.params.fill(params)  # fill overwrites old params
+
+        # Fill locations tab
+        if locations:
+            view.locations.multiselect.set_assigned(locations)
+
+        # Fill organizations tab
+        if organizations:
+            view.organizations.multiselect.set_assigned(organizations)
+
+        view.submit_button.click()
+
+    def create(self, dns_domain, **kwargs):
+        """
+        Create a new domain.
+
+        :param dns_domain: str, full DNS domain name
+        :param full_name: (optional) str, full name describing the domain
+        :param dns_capsule: (optional) str of DNS capsule to use
+        :param parameters: (optional) param dict or list of param dicts to set
+            Format: [{'name': str, 'value': str}]
+        :param locations: list of location strings to be assigned.
+            All other locations will be unassigned.
+            Default: ['Default Location']
+        :param organizations: list of org strings to be assigned.
+            All other orgs will be unassigned.
+            Default: ['Default Organization']
+        """
+        view = self.navigate_to(self, 'New')
+
+        kwargs['dns_domain'] = dns_domain
+
+        if not kwargs.get('locations'):
+            kwargs['locations'] = [self._DEFAULT_LOC]
+        if not kwargs.get('organizations'):
+            kwargs['organizations'] = [self._DEFAULT_ORG]
+
+        self._submit_domain_form(view, **kwargs)
+
+        wait_for(
+            lambda: view.flash.messages, fail_condition=[], timeout=10,
+            delay=2, message='wait for flash success'
+        )
+        view.flash.assert_success_message(self._CREATE_SUCCESSFUL.format(dns_domain))
+
+    def search(self, value=None):
```

Comment:
we already have `SearchableViewMixin` that should handle all the logic that we need, so usually we just use:

```
        view = self.navigate_to(self, 'All')
        return view.search(value, expected_result)
```

Reply:
Right, I noticed a few flaws with that Mixin so I extended this method slightly. For example, if you tried to do a `view.search('a')` you'd get back the first text on the page that has the character `a` on it. This would be `Default Organization`. I do realize you can change the `search_result_locator`, but what if you have multiple entries with 'a' in their name? So I decided to have it iterate through the rows on the search table and return a list of all matches that search has returned.

If every page that we use `SearchableViewMixin` has a table on it I could move this logic into there... but we'd have to be careful if tests are expecting to get 1 string back instead of a list.

Reply:
That's what i'm currently working on - updating that mixing to return table values, not just string by locator. Pls postpone this comment for now and resolve others - i hope i'll submit PR soon so you'll just update yours to use updated mixin.